File: backend/app/schemas/query.py
"""Pydantic schemas for self-service query endpoint."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class QueryRunRequest(BaseModel):
    """Request body for POST /query/run."""

    sql: str = Field(
        ...,
        min_length=10,
        max_length=50_000,
        description="SELECT-only SQL query to execute against the event store.",
    )
    org: str | None = Field(
        None,
        max_length=255,
        pattern=r"^[a-zA-Z0-9_.-]+$",
        description="Optional org to narrow query scope.",
    )
    format: str = Field(
        default="json",
        pattern=r"^(json|csv)$",
    )


class QueryRunResponse(BaseModel):
    """Response for POST /query/run."""

    columns: list[str]
    rows: list[list[Any]]
    row_count: int
    truncated: bool
    execution_ms: int
    query_id: str


class QueryTemplate(BaseModel):
    """Saved query template."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None = None
    sql: str
    created_by: str | None = None
    org_slug: str | None = None
    created_at: datetime
    updated_at: datetime | None = None


class QueryTemplateCreate(BaseModel):
    """Request body for creating a query template."""

    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = Field(None, max_length=1000)
    sql: str = Field(..., min_length=10, max_length=50_000)
    org_slug: str | None = Field(None, max_length=255)


# ── Saved Queries ─────────────────────────────────────────────────────────────


class SavedQueryCreate(BaseModel):
    """Request body for saving a query."""

    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = Field(None, max_length=2000)
    sql_text: str = Field(..., min_length=10, max_length=50_000)
    tags: list[str] | None = Field(None, max_length=20)


class SavedQueryUpdate(BaseModel):
    """Request body for updating a saved query."""

    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = Field(None, max_length=2000)
    sql_text: str | None = Field(None, min_length=10, max_length=50_000)
    tags: list[str] | None = None


class SavedQueryResponse(BaseModel):
    """Response schema for a saved query."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None = None
    sql_text: str
    owner_login: str
    is_shared: bool = False
    shared_with: list[str] | None = None
    tags: list[str] | None = None
    schedule_cron: str | None = None
    schedule_enabled: bool = False
    last_run_at: datetime | None = None
    created_at: datetime
    updated_at: datetime | None = None


class ShareQueryRequest(BaseModel):
    """Request body for sharing a query with other users."""

    logins: list[str] = Field(..., min_length=1, max_length=50)


class ScheduleQueryRequest(BaseModel):
    """Request body for scheduling a saved query."""

    cron: str = Field(..., min_length=5, max_length=100)
    enabled: bool = True


class SchemaColumn(BaseModel):
    """A column in a database table schema."""

    name: str
    type: str


class SchemaTable(BaseModel):
    """A table in the database schema with its columns."""

    table: str
    columns: list[SchemaColumn]

File: frontend/src/types/query.ts
export interface QueryRunRequest {
  sql: string;
  org?: string;
  format?: 'json' | 'csv';
}

export interface QueryRunResponse {
  readonly columns: readonly string[];
  readonly rows: readonly (readonly unknown[])[];
  readonly row_count: number;
  readonly truncated: boolean;
  readonly execution_ms: number;
  readonly query_id: string;
}

export interface QueryTemplate {
  readonly id: number;
  readonly name: string;
  readonly description: string | null;
  readonly sql: string;
  readonly created_by: string;
  readonly created_at: string;
}

export interface QueryTemplateCreate {
  name: string;
  description?: string;
  sql: string;
}

// ── Saved Queries ────────────────────────────────────────────────────────────

export interface SavedQuery {
  readonly id: number;
  readonly name: string;
  readonly description: string | null;
  readonly sql_text: string;
  readonly owner_login: string;
  readonly is_shared: boolean;
  readonly shared_with: readonly string[] | null;
  readonly tags: readonly string[] | null;
  readonly schedule_cron: string | null;
  readonly schedule_enabled: boolean;
  readonly last_run_at: string | null;
  readonly created_at: string;
  readonly updated_at: string | null;
}

export interface SavedQueryCreate {
  name: string;
  description?: string;
  sql_text: string;
  tags?: string[];
}

export interface SavedQueryUpdate {
  name?: string;
  description?: string;
  sql_text?: string;
  tags?: string[];
}

// ── Schema ──────────────────────────────────────────────────────────────────

export interface SchemaColumn {
  readonly name: string;
  readonly type: string;
}

export interface SchemaTable {
  readonly table: string;
  readonly columns: readonly SchemaColumn[];
}
